### PR TITLE
cloudinfo: Adding AWS, Azure support and InstanceID

### PR DESCRIPTION
**Pull request by enoodle**

### CODE REVIEW

**Comment by enoodle** [comment]
what is the Authorized Contributors Group for the CLA ?

**Comment by enoodle** [comment]
@simon3z

**Comment by vishh** [comment]
ok to test

**Comment by vishh** [comment]
Overall lgtm

File: manager/machine.go [hunk 1]
```diff
@@ -96,6 +96,7 @@ func getMachineInfo(sysFs sysfs.SysFs, fsInfo fs.FsInfo, inHostNamespace bool) (
 	realCloudInfo := cloudinfo.NewRealCloudInfo()
 	cloudProvider := realCloudInfo.GetCloudProvider()
 	instanceType := realCloudInfo.GetInstanceType()
+	instanceID := realCloudInfo.GetInstanceID()
```
Comment (simon3z) [inline manager/machine.go]: Can this take a long time? (e.g. is it polling ip addresses that may not be always available?)
Would that be a problem? Should this be async?

File: utils/cloudinfo/cloudinfo.go [hunk 2]
```diff
@@ -23,19 +23,23 @@ import (
...
 type realCloudInfo struct {
 	cloudProvider info.CloudProvider
 	instanceType  info.InstanceType
+	instanceID    info.InstanceID
 }
 
 func NewRealCloudInfo() CloudInfo {
 	cloudProvider := detectCloudProvider()
 	instanceType := detectInstanceType(cloudProvider)
+	instanceID := detectInstanceID(cloudProvider)
```
Comment (simon3z) [inline utils/cloudinfo/cloudinfo.go]: Can this take a long time? (e.g. is it polling ip addresses that may not be always available?)
Would that be a problem? Should this be async?

Comment (enoodle) [inline utils/cloudinfo/cloudinfo.go]: I think detectCloudProvider could take some time. The others will be almost immediate as we are accessing the correct API server. In short this function could block for some time and its users should be aware of this as it was already blocking before this patch.

Comment (simon3z) [inline utils/cloudinfo/cloudinfo.go]: @enoodle ok :+1:

File: utils/cloudinfo/azure.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	BiosUUIDFileName     = "/sys/class/dmi/id/product_uuid"
+	MicrosoftCorporation = "Microsoft Corporation"
+)
+
+func onAzure() bool {
+	data, err := ioutil.ReadFile(SysVendorFileName)
+	if err != nil {
+		return false
+	}
+	return strings.Contains(string(data), MicrosoftCorporation)
```
Comment (simon3z) [inline utils/cloudinfo/azure.go]: @enoodle is this a stable check? Is it documented?

Comment (enoodle) [inline utils/cloudinfo/azure.go]: No. I couldn't find any api service comparable to aws/gce 's for Azure, nor any other official documentation on ways to detect if an instance is running on Azure. I did find [1] which explains how to use dmidecode to get the instance ID (hence /sys/class/dmi/id/produect_uuid) and I noticed that on all my instances I had that string for sys_vendor so I used this.

[1] https://azure.microsoft.com/en-us/blog/accessing-and-using-azure-vm-unique-id/

**Comment by enoodle** [comment]
@deads2k can you review this? Thanks

**Comment by deads2k** [comment]
@googlebot @enoodle is covered under the Red Hat CLA.

**Comment by deads2k** [comment]
@enoodle When you add the godep packages, can you name the commit `bump(github.com/aws/aws-sdk-go) 72440a9c5a884936f1c679591fcf04e31a9b772a` (that looks like what you were trying to get).  If the rest are just transitive that's fine, but that name will help track back to the target.

**Comment by deads2k** [comment]
Is this still a WIP?  If so, what is remaining?

**Comment by enoodle** [comment]
@deads2k Nothing more to implement. It was wip for the dependencies.

**Comment by enoodle** [comment]
@deads2k I will rebase to change the dependencies commit name in a moment.

**Comment by deads2k** [comment]
@vishh Implementation is done.  This enables fallback behavior for `ProviderID` determination in the kubelet (see https://github.com/kubernetes/kubernetes/pull/21373).  Take another look?

**Comment by deads2k** [comment]
@jimmidyson ?

**Comment by jimmidyson** [comment]
@deads2k Sorry on vacation.

@vishh Could you review when you get a chance please?

**Comment by enoodle** [comment]
@vishh ping? you actually lgtm'ed it before, I only added a tiny fix to remove \n char when reading the azure instance ID from file.

**Comment by vishh** [comment]
Sorry. This somehow fell off my radar. LGTM again

**Pull request merged**